object/slab.go

```diff
@@ -72,26 +72,43 @@ func (s Slab) Encode(buf []byte, shards [][]byte) {
 	}
 }
 
-// Reconstruct reconstructs the missing shards of a slab. Missing shards must
-// have a len of zero. All shards should have a capacity of at least
-// rhpv2.SectorSize, or they will be reallocated.
-func (s Slab) Reconstruct(shards [][]byte) error {
+// ReconstructSome reconstructs the required shards of a slab.
+func (s Slab) ReconstructSome(shards [][]byte, required []bool) error {
 	for i := range shards {
+		// Make sure shards are either empty or full.
 		if len(shards[i]) != rhpv2.SectorSize && len(shards[i]) != 0 {
 			panic("shards must have a len of either 0 or rhpv2.SectorSize")
 		}
-		if cap(shards[i]) < rhpv2.SectorSize {
-			shards[i] = make([]byte, 0, rhpv2.SectorSize)
-		}
-		if len(shards[i]) != 0 {
-			shards[i] = shards[i][:rhpv2.SectorSize]
+		// Every required shard needs to have a sector worth of capacity.
+		if required[i] && cap(shards[i]) < rhpv2.SectorSize {
+			shards[i] = reedsolomon.AllocAligned(1, rhpv2.SectorSize)[0][:0]
```

Comment:
it's worth adding a test case for this if there isn't one already

Reply:
We don't have a unit test but this is hit by `TestMigrations` since we don't preallocate the required shards when downloading.